### PR TITLE
Add big warning about averaging to KLDivLoss documentation #6622

### PR DESCRIPTION
**Pull request by t-vi**
Also: An (independent) change in `torch.diagonal` docstring tensor
formatting.

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -290,6 +290,11 @@ class KLDivLoss(_Loss):
     .. _Kullback-Leibler divergence:
         https://en.wikipedia.org/wiki/Kullback-Leibler_divergence
 
+    .. note:: The default averaging means that the loss is actually **not** the
```
Comment (zou3519) [inline torch/nn/modules/loss.py]: Nit: you should mention `size_average` and tell the users how to get the correct KL div behavior.
Something like:

To get the real KL Divergence, use the ``size_average=False``, then divide the output of the module by the number of minibatches.

Example:
   loss = nn.KLDiv(size_average=False)
   out = loss(tensor1, tensor2) / tensor1.size(0)

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -290,6 +290,19 @@ class KLDivLoss(_Loss):
...
+          KL Divergence because the terms are already probability weighted.
+          A future release of PyTorch may move the default loss closer to the
+          mathematical definition.
+
+          To get the real KL Divergence, use `size_average=False`, and
+          then divide the output by the batch size.
+
+          Example::
+
+              loss = nn.KLDiv(size_average=False)
```
Comment (zou3519) [inline torch/nn/modules/loss.py]: My bad, it's called nn.KLDivLoss

File: torch/nn/modules/loss.py [hunk 3]
```diff
@@ -290,6 +290,19 @@ class KLDivLoss(_Loss):
     .. _Kullback-Leibler divergence:
         https://en.wikipedia.org/wiki/Kullback-Leibler_divergence
 
+    .. note:: The default averaging means that the loss is actually **not** the
+          KL Divergence because the terms are already probability weighted.
+          A future release of PyTorch may move the default loss closer to the
+          mathematical definition.
+
+          To get the real KL Divergence, use `size_average=False`, and
+          then divide the output by the batch size.
```
Comment (apaszke) [inline torch/nn/modules/loss.py]: Can we please get a code example for this? Also it’s not just loss / batch_size, it’s loss.sum() / batch_size

Comment (fmassa) [reply torch/nn/modules/loss.py]: `size_average=False` already performs the reduction, so no need to use the `.sum()`

Comment (t-vi) [reply torch/nn/modules/loss.py]: So I updated the example to have `batch_size` and actual output.

Comment (apaszke) [reply torch/nn/modules/loss.py]: Uh I confused `size_average` with `reduce`! Thanks @fmassa

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -285,11 +285,29 @@ class KLDivLoss(_Loss):
...
+          mathematical definition.
+
+          To get the real KL Divergence, use ``size_average=False``, and
+          then divide the output by the batch size.
+
+          Example::
+
+            >>> loss = nn.KLDivLoss(size_average=False)
+            >>> batch_size = 5
+            >>> probs1 = F.softmax(torch.randn(batch_size, 10), 1)
```
Comment (apaszke) [inline torch/nn/modules/loss.py]: Can we please `log_softmax` here? `log(softmax(x))` is really unstable, and we shouldn't recommend it.

**Review by zou3519 — APPROVED** [review]
lgtm, thanks @t-vi!

**Review by ezyang — APPROVED** [review]
